lib/io/event/selector/select.rb

```diff
@@ -99,52 +100,15 @@ def ready?
 				!@ready.empty?
 			end
 			
-			Waiter = Struct.new(:fiber, :events, :tail) do
-				def alive?
-					self.fiber&.alive?
-				end
-				
-				# Dispatch the given events to the list of waiting fibers. If the fiber was not waiting for the given events, it is reactivated by calling the given block.
-				def dispatch(events, &reactivate)
-					# We capture the tail here, because calling reactivate might modify it:
-					tail = self.tail
-					
-					if fiber = self.fiber
-						if fiber.alive?
-							revents = events & self.events
-							if revents.zero?
-								reactivate.call(self)
-							else
-								self.fiber = nil
-								fiber.transfer(revents)
-							end
-						else
-							self.fiber = nil
-						end
-					end
-					
-					tail&.dispatch(events, &reactivate)
-				end
-				
-				def invalidate
-					self.fiber = nil
-				end
-				
-				def each(&block)
-					if fiber = self.fiber
-						yield fiber, self.events
-					end
-					
-					self.tail&.each(&block)
-				end
-			end
-			
 			def io_wait(fiber, io, events)
-				waiter = @waiting[io] = Waiter.new(fiber, events, @waiting[io])
+				@waiting[io][fiber] = events
 				
 				@loop.transfer
 			ensure
-				waiter&.invalidate
+				if @waiting && @waiting.key?(io)
```

Comment:
Do we need to care about `@waiting` being `nil` here?

If so, what about:

```suggestion
if fibers = @waiting[io] # or @waiting&.dig(io)
  fibers.delete(fiber)
  @waiting.delete(io) if fibers.empty?
end
```

Reply:
Not sure, `@waiting` is set to `nil` in `#close`, so I wonder if the fibers could have to deal with unwinding after that.

And I was using `#key?` there to avoid triggering the Hash's default_proc (which doesn't just return a new hash, but inserts it in it).

Reply:
Ah, that all makes sense. Maybe just `@waiting&.key?(io)` is my only suggestion, but I don't have a strong opinion either way.